vg/vgtex/canvas.go

```diff
@@ -0,0 +1,272 @@
+// Copyright 2016 The gonum Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package vgtex provides a vg.Canvas implementation for LaTeX, using the pgf LaTeX package.
+package vgtex
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"image/color"
+	"io"
+	"math"
+	"strings"
+
+	"github.com/gonum/plot/vg"
+)
+
+const (
+	pr      = 5 // precision to use when outputting float64s
+	rad2deg = 180 / math.Pi
+)
+
+// Canvas implements the vg.Canvas interface, translating gonum/plot
+// drawing primitives into pgf ones.
+type Canvas struct {
+	buf   *bytes.Buffer
+	w, h  vg.Length
+	dpi   int // number of dots per inch for this canvas
+	stack []context
+}
+
+type context struct {
+	color      color.Color
+	dashArray  []vg.Length
+	dashOffset vg.Length
+	linew      vg.Length
+}
+
+// New returns a new LaTeX canvas.
+func New(w, h vg.Length) *Canvas {
+	c := &Canvas{
+		buf: new(bytes.Buffer),
+		w:   w,
+		h:   h,
+		dpi: 72,
+	}
+	c.wtex(`%%%% gonum/plot created for LaTeX/pgf`)
+	c.wtex(`%%%% you need to add:`)
+	c.wtex(`%%%%   \usepackage{pgf}`)
+	c.wtex(`%%%% to your LaTeX document`)
+	c.wtex("")
+	c.wtex(`\begin{pgfpicture}`)
+	c.stack = make([]context, 1)
+	vg.Initialize(c)
+	return c
+}
+
+func (c *Canvas) ctx() *context {
+	return &c.stack[len(c.stack)-1]
+}
+
+func (c *Canvas) Size() (w, h vg.Length) {
+	return c.w, c.h
+}
+
+// SetLineWidth sets the width of stroked paths.
+// If the width is not positive then stroked lines
+// are not drawn.
+//
+// The initial line width is 1 point.
+func (c *Canvas) SetLineWidth(w vg.Length) {
+	c.ctx().linew = w
+}
+
+// SetLineDash sets the dash pattern for lines.
+// The pattern slice specifies the lengths of
+// alternating dashes and gaps, and the offset
+// specifies the distance into the dash pattern
+// to start the dash.
+//
+// The initial dash pattern is a solid line.
+func (c *Canvas) SetLineDash(pattern []vg.Length, offset vg.Length) {
+	c.ctx().dashArray = pattern
+	c.ctx().dashOffset = offset
+}
+
+// SetColor sets the current drawing color.
+// Note that fill color and stroke color are
+// the same so if you want different fill
+// and stroke colors then you must use two
+// separate calls to SetColor.
```

Comment:
This note could be clearer. Presumably you mean you need to set color, draw fills, set color and then draw lines.